### PR TITLE
[WIP] Ansible Builder CLI 'build' Command Implementation

### PR DESCRIPTION
**Pull request by beeankha**
From Issue https://github.com/ansible/ansible-builder/issues/12

### CODE REVIEW

File: ansible_builder/cli.py [hunk 1]
```diff
@@ -8,22 +8,35 @@ def prepare(args=sys.argv[1:]):
...
+
+    for p in [create_command_parser, build_command_parser]:
+
+        p.add_argument('-f', '--file',
+                       default='execution-environment.yml',
+                       dest='filename',
+                       help='The definiton of the execution environment.')
+
+        p.add_argument('-b', '--base-image',
+                       default='shanemcd/ansible-runner',
```
Comment (AlanCoding) [inline ansible_builder/cli.py]: This default is written twice now, in the other file

```python
default_base_image = 'shanemcd/ansible-runner'
```

I would like to consolidate them. We could do that we a constants file.

**Comment by AlanCoding** [comment]
I put up suggestions in this pull request

https://github.com/beeankha/ansible-builder/pull/1

But you can just push the commit to avoid a merge commit. I will go over usage now

----

```
$ ansible-builder create -c alan/bc/ -f alan/alan_ee.yml --container-runtime=docker
Complete! Build context is at: alan/bc/
$ ansible-builder build -c alan/bc/ -f alan/alan_ee.yml --container-runtime=docker
Complete! Build context is at: alan/bc/
$ docker images
REPOSITORY                                   TAG                 IMAGE ID            CREATED             SIZE
ansible-execution-env                        latest              97bc9698bd0c        13 seconds ago      628MB
shanemcd/ansible-runner                      latest              076625a99a3b        24 hours ago        628MB
gcr.io/ansible-tower-engineering/awx_devel   devel               15b69da6ccdf        2 weeks ago         1.83GB
memcached                                    alpine              a8da907c7f84        3 months ago        9.22MB
redis                                        latest              f0453552d7f2        3 months ago        98.2MB
postgres                                     10                  538942384eaf        3 months ago        200MB
```

Existence of "ansible-execution-env" shows that this is finishing.

That's with the setup:

```
$ cat alan/alan_ee.yml 
---
# subset of what may eventually be the container to replace the Ansible venv in AWX
version: 1
dependencies:
  galaxy: alan/scm_metadata.yml  # TODO: remove alan folder later
  python: requirements.txt
  system:
    - git  # for project updates
 
additional_build_steps: |
  RUN some-command
  ENV SOME_VAR=some_value
$ cat alan/scm_metadata.yml 
collections:
   - name: https://github.com/alancoding/azure.git
     version: ee
     type: git
# will fill in the rest of them as branches are created
```

----

Comments that go beyond what I have in the PR:

 - the command names are not very obvious, it would be hard to guess what "create" vs "build" does.
 - no help text for these commands to find out.
 - the build command doesn't give any output that indicates what it did.
 - it's confusing that the build command takes both the `-f` and the `-c` options. Is `-c` the input, or an output? Maybe the build context should be in a tmpdir it's an intermediary step.

File: ansible_builder/main.py [hunk 2]
```diff
@@ -1,19 +1,29 @@
...
-        self.build_context = build_context or os.path.join(os.getcwd(), 'context')
+        self.tag = tag
+        self.build_context = build_context
+        self.container_runtime = container_runtime
         self.containerfile = Containerfile(
-            filename='Containerfile',
+            filename={
+                'podman': 'Containerfile',
+                'docker': 'Dockerfile'
+            }[self.container_runtime],
```
Comment (shanemcd) [inline ansible_builder/main.py:27]: This is going to throw a key error if passed anything other than the exact strings "podman" or "docker". If we want to go this route, let's do some validation to make sure this is one or the other. Though it _may_ be desirable to have this be an arbitrary string if you have a wrapper script or something else that you use to invoke these tools, maybe some other use cases I am not thinking of.

File: ansible_builder/main.py [hunk 3]
```diff
@@ -22,9 +32,21 @@ def __init__(self, filename='execution-environment.yml', base_image=default_base
...
 
+    def build(self):
+        self.create()
+        command = [self.container_runtime, "build"]
+        arguments = ["-f", self.containerfile.path,
+                     "-t", self.tag,
+                     self.build_context]
+        for arg in arguments:
+            command.append(arg)
+        result = subprocess.run(command, capture_output=True)
```
Comment (shanemcd) [inline ansible_builder/main.py:51]: I would love to have a test that asserts that `command` contains certain elements, whether or not we override things like the build context path or the tag. We shouldn't actually be invoking `docker build` in our unit tests, so we'll want to move the command construction into a separate method that can be tested in isolation.

File: ansible_builder/main.py (hunk 2 above)
Comment (AlanCoding) [reply ansible_builder/main.py:27]: Yes, the argparse definition of the option should have these two choices. We can put the mapping inside the constants, and also define the options by making a list of the keys of this dict.

File: ansible_builder/cli.py [hunk 4]
```diff
@@ -2,28 +2,44 @@
...
+                       default=constants.default_file,
+                       dest='filename',
+                       help='The definiton of the execution environment.')
+
+        p.add_argument('-b', '--base-image',
+                       default=constants.default_base_image,
+                       help='The parent image for the execution environment.')
+
+        p.add_argument('-c', '--context',
+                       default=None,
```
Comment (AlanCoding) [inline ansible_builder/cli.py]: looks like the test failures are due to this lacking a default, which you could put in `constants.default_build_context` for

File: ansible_builder/cli.py [hunk 5]
```diff
@@ -34,8 +50,10 @@ def run():
     ab = prepare()
 
     print('Processing...', end='\r')
+    # TO DO: stdout that prints in real-time and/or a "spinner"
 
-    if ab.process():
+    build_or_create = getattr(ab, sys.argv[1])
```
Comment (AlanCoding) [inline ansible_builder/cli.py]: this use of `sys.argv` undermines the role of argparse.

The only way to solve this is to make the action one of the primary parameters in this main data bus. Consider:

```diff
diff --git a/ansible_builder/cli.py b/ansible_builder/cli.py
index 77b1571..eda4f36 100644
--- a/ansible_builder/cli.py
+++ b/ansible_builder/cli.py
@@ -7,7 +7,7 @@ from . import constants
 
 def prepare(args=sys.argv[1:]):
     parser = argparse.ArgumentParser(prog='ansible-builder')
-    subparsers = parser.add_subparsers(help='The command to invoke.')
+    subparsers = parser.add_subparsers(help='The command to invoke.', dest='action')
 
     create_command_parser = subparsers.add_parser('create',
                                                   help='Outputs a build context, including a Containerfile, populated with dependencies.')
@@ -33,7 +33,7 @@ def prepare(args=sys.argv[1:]):
                        help='The parent image for the execution environment.')
 
         p.add_argument('-c', '--context',
-                       default=None,
+                       default=constants.default_build_context,
                        dest='build_context',
                        help='The directory to use for the build context. Defaults to $PWD/context.')
 
@@ -52,7 +52,7 @@ def run():
     print('Processing...', end='\r')
     # TO DO: stdout that prints in real-time and/or a "spinner"
 
-    build_or_create = getattr(ab, sys.argv[1])
+    build_or_create = getattr(ab, ab.action)
     if build_or_create():
         print("Complete! Build context is at: {}".format(ab.build_context))
         sys.exit(0)
diff --git a/ansible_builder/main.py b/ansible_builder/main.py
index 33e1285..bcc9068 100644
--- a/ansible_builder/main.py
+++ b/ansible_builder/main.py
@@ -8,12 +8,13 @@ from . import constants
 
 
 class AnsibleBuilder:
-    def __init__(self,
+    def __init__(self, action=None,
                  filename=constants.default_file,
                  base_image=constants.default_base_image,
                  build_context=constants.default_build_context,
                  tag=constants.default_tag,
                  container_runtime=constants.default_container_runtime):
+        self.action = action
         self.definition = Definition(filename=filename)
         self.base_image = base_image
         self.tag = tag
@@ -78,7 +79,10 @@ class Definition:
 class Containerfile:
     newline_char = '\n'
 
-    def __init__(self, *args, filename, definition, build_context, base_image):
+    def __init__(self, definition,
+                 filename=constants.default_file,
+                 build_context=constants.default_build_context,
+                 base_image=constants.default_base_image):
         self.build_context = build_context
         os.makedirs(self.build_context, exist_ok=True)
         self.definition = definition
```

**Comment by AlanCoding** [comment]
Followup on my prior steps, closing loop. This uses syntax stolen from https://github.com/ansible/ansible-runner/pull/470, so it's a little similar to what @jladdjr will be doing with these images.

```
$ docker run --rm --tty --interactive ansible-execution-env ansible-doc -t inventory azure.azcollection.azure_rm
ERROR! inventory azure.azcollection.azure_rm missing documentation (or could not parse documentation): unknown doc_fragment(s) in file /usr/share/ansible/collections/ansible_collections/azure/azcollection/plugins/inventory/azure_rm.py: azure
```

This is expected, I merged a PR that fixes this doc_fragments issue since the last release.

File: ansible_builder/cli.py [hunk 6]
```diff
@@ -34,8 +52,10 @@ def run():
     ab = prepare()
 
     print('Processing...', end='\r')
+    # TO DO: stdout that prints in real-time and/or a "spinner"
```
Comment (shanemcd) [inline ansible_builder/cli.py]: Can we make these all `TODO`? A lot of editors look for this token, and I dont think they are smart enough to find "TO DO" as separate words.

File: ansible_builder/main.py [hunk 7]
```diff
@@ -22,9 +33,25 @@ def __init__(self, filename='execution-environment.yml', base_image=default_base
...
         return self.containerfile.write()
 
+    def build_command(self):
+        self.create()
+        command = [self.container_runtime, "build"]
+        arguments = ["-f", self.containerfile.path,
+                     "-t", self.tag,
+                     self.build_context]
+        for arg in arguments:
+            command.append(arg)
```
Comment (AlanCoding) [inline ansible_builder/main.py:46]: you really shouldn't loop for this, `command.extend(arguments)` would be better

**Review by AlanCoding — APPROVED** [review]

**Pull request merged**